### ISSUE TITLE
SanityAsset : missing properties ?

### ISSUE THREAD

**Issue by ErnestBrandi** [ricokahler/sanity-codegen#165]
Hi, I have a schema with the following part representing an array of images 
```
{
      title: "Images",
      name: "images",
      type: "array",
      of: [{ type: "image" }],
}
```
generated type is the following 
```
images?: Array<
    SanityKeyed<{
      _type: "image";
      asset: SanityAsset;
      crop?: SanityImageCrop;
      hotspot?: SanityImageHotspot;
    }>
  >;
```
with SanityAsset being
```
export declare type SanityAsset = SanityReference<any>;
```
I feel like I miss something but where are all the image properties such as url, path, etc ?

**Comment by ricokahler** [comment]
Hi @ErnestBrandi 👋

ah, thanks for catching this. I think the correct types should've been:

```
  images?: Array<
    SanityKeyed<{
      _type: "image";
      asset: SanityReference<SanityAsset>;
      crop?: SanityImageCrop;
      hotspot?: SanityImageHotspot;
    }>
  >;
```

and then `SanityAsset` should've been the thing with all the correct types. 

e.g.

```ts
interface SanityImageAsset extends SanityDocument {
  _type: 'sanity.imageAsset'
  assetId: string
  extension: string
  metadata: SanityImageMetadata
  mimeType: string
  originalFilename: string
  path: string
  sha1hash: string
  size: number
  uploadId: string
  url: string
}
```

Let me get a PR up adding these.

**Comment by ErnestBrandi** [comment]
Hi and thanks for the prompt reply :)
Still, I am not sure about that `SanityReference<?>` (I am fairly new to both Sanity & Typescript though).
For what I understand this should be used on sanity references (https://www.sanity.io/docs/reference-type), for example on this kind of schema : 
```
{
      title: 'foo',
      name: 'bar',
      type: 'reference',
      to: [{type: 'refschema'}]
}
```
which is not what I have with my array of images (or is it implicit ?).
Because of this I still don't have access to `SanityImageAsset` properties. I feel that removing the use `SanityReference` here would be more appropriate (?) :
```
images?: Array<
    SanityKeyed<{
      _type: "image";
      asset: SanityImageAsset;
      crop?: SanityImageCrop;
      hotspot?: SanityImageHotspot;
    }>
  >;
```

**Comment by ricokahler** [comment]
ah i think i know what you mean.

i believe the behavior is working as expected but in your defense, it's not obvious on how this works… I mean the types were wrong prior to you opening this issue, so I got it wrong the first time around haha 😅

---

here's some clarification. let me know if any of this doesn't makes sense:

1. The current version of sanity-codegen only supports generating types from the schema. This means the resulting type shapes will only represent the shape of the raw JSON documents without any expansions or transformations (i.e. the same shape as the result from the [`/data/doc/:documentId` endpoint](https://www.sanity.io/docs/http-doc)). Note that the next version of sanity-codegen is a complete rewrite with many many new features including code generation from [GROQ queries](https://www.sanity.io/docs/how-queries-work) that let you expand references and reshape data.<br/>
This is the reason why you're seeing your image typed as `asset: SanityReference<SanityImageAsset>;`, because the raw document will not actually include that information.<br/>
*However, note that you don't need to resolve the asset in order to get a URL because the ID of the asset is all you need to generate a URL. see [here](https://www.sanity.io/docs/image-urls) and [here](https://www.sanity.io/docs/image-url))

2. A sanity image type (i.e. `type: 'image'`) is actually an extension of the `'object'` type under the hood that provides a default field called `asset`. Under the hood, this `asset` field uses the same mechanism as sanity references and is even of `type: 'reference'` when looking at the raw document data.<br/>
In order to get all the information listed in `SanityImageAsset`, you need to expand this reference. This could be done by querying for the referenced document by its ID or by using the [GROQ reference access operator](https://www.sanity.io/docs/groq-operators#dae298fc7952) `asset->`.

```ts
import sanityClient from '@sanity/client';
import * as Schema from './your-generated-types';

const exampleDocument = {
  _id: '875c8b70-ebc0-4fb5-81fd-07d125693bdd',
  _type: 'blogPost',
  // ...
  // rest of document fields
  // ...
  
  // this is the shape of a raw image field without any expansions
  myImage: {
    asset: {
      // notice how assets are references on the document when queried without expansions
      '_type': 'reference',
      '_ref': 'image-69ad5d60ff19c456954513e8c67e9563c780d5e1-780x1170-jpg',
    },
  },
  // ...
  // rest of document fields
  // ...
}

const client = sanityClieint({
  // see here for sanity client docs:
  // https://www.sanity.io/docs/js-client
});

// this is a groq query being made by the official sanity client.
// this query queries for the whole document without expansions and will
// match the generated type
//
const exampleDocAgain = await client.fetch<Schema.BlogPost>(
  //                                                  👆
  // the first-party client provides a place for type parameter
  // which can be filled in with the generated schema types
  `*[_id == $id][0]`,
  {id: '875c8b70-ebc0-4fb5-81fd-07d125693bdd'}
);

// this is a query expands the image asset using the reference access operator `->`
const expandedExampleDoc = await client.fetch<BlogPostWithExpandedImage>(
  `*[_id == $id][0] {
     // use the splat to get the rest of the properties unaltered
     ...,
     // replace the 'myImage' key with this projection that expands the asset reference
     myImage: { asset-> },
   }`
)

// the current version of sanity-codegen does not support autogenerating types
// from groq so you have to build up your altered shape yourself. 
type BlogPostWithExpandedImage = Omit<Schema.BlogPost, 'myImage'> & {
  myImage: { asset: Schema.SanityImageAsset }
}

// alternatively, you could query for the ID of the referenced document directly
// instead of expanding it in GROQ. the ID to query is the `_ref` of the reference
// object.
const assetDetails = await client.fetch<Schema.SanityImageAsset>(
  `*[_id == $ref][0]`,
  {ref: 'image-69ad5d60ff19c456954513e8c67e9563c780d5e1-780x1170-jpg'}
);
//   👆
// this will return `{ url: '', metadata: {}, path: [], }` etc
```

----

TL;DR, assets are sanity references under the hood and you need to resolve in order to get the info you want out of them. The most common way people do this is via GROQ, however the current version of sanity codegen doesn't support code generation from GROQ. For now you can utilize the generated schema types paired with type literals and [utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) to build up the shape you want.

Admittedly, this is still a somewhat manual process but I'm planning on releasing that beta with groq-codegen [soon<sup>TM</sup>](https://github.com/ricokahler/sanity-codegen/issues/5#issuecomment-877925182). When that's done, you don't have to write any types yourself.

Additionally, check out the [official image builder package](https://www.sanity.io/docs/image-url) for generating optimized URLs dynamically. This package accepts an image type without needing to resolve asset at all  (via [image URLs](https://www.sanity.io/docs/image-urls)).

**Comment by ErnestBrandi** [comment]
Woah thanks for that very detailed explanation !

However my case it a bit more specific. First I do not use GROQ but Graphql myself (which seems to alter the way data is represented). 
So I do not have any trouble getting all the data I want from the asset object (`url` for instance) but on the other hand the `_ref` property doesn't exists... So there is no possible match between the `_ref` property exposed by the typescript interface and the actual data I can possibly receive from the graphql endpoint (and therefore I cannot use the image builder package as you suggest).
I understand the package as no goal representing data which is not explicitly represented in the schema and that an image asset is under the hood just a reference; but again does it worth keeping that abstraction up ? Why not just defining the asset as an `SanityImageAsset` ? 
For now this is the solution I am using in order to access assets properties nevertheless :  `<Image src={(mainImage.asset as unknown as SanityImageAsset).url} />`

**Comment by ricokahler** [comment]
> First I do not use GROQ but Graphql myself

ah then i have a simple answer for you: you're using the wrong project lol

if you haven't heard of it yet, checkout [grapgql-codegen](https://www.graphql-code-generator.com)

it's a generic project unrelated to sanity that works with any graphql endpoint. they use the introspection results to convert the types from graphql to typescript.

i should definitely include this tip in the readme though

**Comment by ErnestBrandi** [comment]
Uh I have even thought about relying on Graphql schemas only (as I don't write them myself)... That does indeed solve my issue :) Thanks a lot for your time and keep up the good work !

**Comment by genox** [comment]
I resolve all references within groq queries already and therefore expect object properties to be there. With a large number of references and nested items in many document types, I was looking for a bit of a hack to get this to work nicely. I ended up with this:

```
// import instead of sanity-codegen provided types
// import { SanityKeyedReference, SanityReference } from 'types/api/sanityCodegenMod';
export declare type SanityReference<T> = T & {
  _ref: string;
};

export declare type SanityKeyedReference<T> = T & {
  _key: string;
  _ref: string;
};
```

Importing these types on top of the generated types and removing the existing imports allows me to have all properties available without having to augment/omit any existing sanity object or document properties.
